### ISSUE TITLE
Unit test CI build failure in pkg test

### ISSUE THREAD

**Issue by emlin** [vmware/vic#2329]
see https://ci.vmware.run/vmware/vic/4784
# github.com/vmware/vic/vendor/github.com/vmware/govmomi/object

vendor/github.com/vmware/govmomi/object/host_service_system.go:25: can't find import: "github.com/vmware/vic/vendor/github.com/vmware/govmomi/vim25/types"
Makefile:227: recipe for target 'test-job-github.com/vmware/vic/pkg' failed
make: **\* [test-job-github.com/vmware/vic/pkg] Error 2
make: **\* Waiting for unfinished jobs....

**Comment by emlin** [comment]
And didn't see any file imported github.com/vmware/vic/vendor/github.com/vmware/govmomi/vim25/types

**Comment by hickeng** [comment]
You're looking for an import of `github.com/vmware/govmomi/vim25/types` - Go should be honoring the vendor directory as of 1.5 experimental iirc.

**Comment by mhagen-vmware** [comment]
https://ci.vmware.run/vmware/vic/4906

**Comment by gigawhitlocks** [comment]
Swiping this one from you @emlin

**Comment by gigawhitlocks** [comment]
This is really strange; I was able to reproduce this exactly once (though actually it failed on a different package)...

**Comment by gigawhitlocks** [comment]
Quite a heisenbug. Doesn't reproduce in a clean golang container on the troubled commit or `master`, or in a golang container w/ a volume mount to source & possibly artifacts from previous builds; doesn't reproduce before or after a clean build. 

We should probably just close this and reopen it if the same exact error is seen again and/or the issue becomes reproducible.

**Comment by emlin** [comment]
> This is really strange; I was able to reproduce this exactly once (though actually it failed on a different package)...

I could understand such kind of issue probably happen anywhere, that just not error inside of our code base. It's but in either docker or go test

**Comment by mhagen-vmware** [comment]
Eh, it has already happened three times. (Once for an Ian build that you restarted)  It is a challenging race condition but it keeps happening.  I definitely don't want to do nothing here and just close it. Again, is there more logging at least that we can add?

**Comment by gigawhitlocks** [comment]
Reopened at @mdubya66 's request;

Also this
>  It is a challenging race condition but it keeps happening
is not what's happening here. There's no parallelism at play that could be causing a race at this point in the build process. It's transient, in that we've seen it in 3 builds out of many at this point, and in my opinion is likely a bug w/in the Go coverage checker in the `go` tool itself, or occurs if something causes `git` to fail to clone the entire branch before the build.

**Comment by andrewtchin** [comment]
seen again in https://ci.vmware.run/vmware/vic/5059
as a hack can we wrap the coverage in a retry?

**Comment by dougm** [comment]
@gigawhitlocks There is parallelism at this point, we are using `make -j3 ...`: https://github.com/vmware/vic/blob/d8461eabd5634fc826582a28ad7630b95724230f/.drone.yml#L26

**Comment by dougm** [comment]
Does it always fail in govmomi/vim25/types?  That package is huge, so huge the Go team uses it for their compiler benchmarks.  I don't know how the cover tool(s) work, but it might be worthwhile to make sure the govmomi package is compiled (installed) before they run.  As in:
```
go install ./vendor/github.com/vmware/govmomi
```

This will put the compiled archives in $GOPATH/pkg, so the cover tool wouldn't have to compile it in $TMPDIR everytime - if that's what its currently doing.

**Comment by mhagen-vmware** [comment]
@dougm Yes, every time this type of error has happened it is in that specific package.

**Comment by anchal-agrawal** [comment]
https://ci.vmware.run/vmware/vic/5207

**Comment by gigawhitlocks** [comment]
> That package is huge, so huge the Go team uses it for their compiler benchmarks

Incredible. It's also so huge that my tooling chokes on it in emacs when I open the file. Every time. Gocode just breaks on it. I think we ought to do what @dougm is suggesting

**Comment by anchal-agrawal** [comment]
https://ci.vmware.run/vmware/vic/5308

**Comment by gigawhitlocks** [comment]
> @gigawhitlocks There is parallelism at this point, we are using make -j3 ...: https://github.com/vmware/vic/blob/d8461eabd5634fc826582a28ad7630b95724230f/.drone.yml#L26

Oh I just saw this / didn't realize it before. This could be why I was unable to replicate it, as I was calling `make` w/o a `-j` argument while attempting to do so

**Comment by gigawhitlocks** [comment]
OK I'm gonna go ahead and do what @dougm was suggesting while my other no-repro bug bastes in its newfound debugging info

**Comment by gigawhitlocks** [comment]
#2601 tentatively fixes this. Hard to confirm w/ no-repro but I'll close this issue at least for the time being but hopefully for good once I merge that PR.

**Comment by hickeng** [comment]
https://ci.vmware.run/vmware/vic/5528 is failing with this:
```
# github.com/vmware/vic/vendor/github.com/vmware/govmomi/vim25/mo
vendor/github.com/vmware/govmomi/vim25/mo/retrieve.go:23: can't find import: "github.com/vmware/vic/vendor/github.com/vmware/govmomi/vim25/methods"
Makefile:227: recipe for target 'test-job-github.com/vmware/vic/lib' failed
make: *** [test-job-github.com/vmware/vic/lib] Error 2
make: *** Waiting for unfinished jobs....
```

**Comment by hickeng** [comment]
Moving back to in-progress as it's still occurring after `go test -i` merge

**Comment by hickeng** [comment]
https://ci.vmware.run/vmware/vic/5591 (but now restarted)
```
# github.com/vmware/vic/vendor/github.com/vmware/govmomi/vim25/soap
vendor/github.com/vmware/govmomi/vim25/soap/client.go:37: can't find import: "github.com/vmware/vic/vendor/github.com/vmware/govmomi/vim25/types"
Makefile:227: recipe for target 'test-job-github.com/vmware/vic/lib' failed
make: *** [test-job-github.com/vmware/vic/lib] Error 2
make: *** Waiting for unfinished jobs....
```

**Comment by mhagen-vmware** [comment]
https://ci.vmware.run/vmware/vic/6613

**Comment by mhagen-vmware** [comment]
https://ci.vmware.run/vmware/vic/6801

**Comment by mhagen-vmware** [comment]
This time we got the output of ls on the directory and yet it still reporting unable to find the import...

**Comment by anchal-agrawal** [comment]
Also seen in https://ci.vmware.run/vmware/vic/9286:

```
Running unit tests
# test everything but vendor
Generating coverage data
...
ok  	github.com/vmware/vic/pkg/vsphere/toolbox	0.010s	coverage: 98.7% of statements
ok  	github.com/vmware/vic/pkg/vsphere/vm	0.111s	coverage: 32.4% of statements
ok  	github.com/vmware/vic/cmd/vicadmin	0.027s	coverage: 2.9% of statements
cover: line [] doesn't match expected format: ^(.+):([0-9]+).([0-9]+),([0-9]+).([0-9]+) ([0-9]+) ([0-9]+)$
cover: line [] doesn't match expected format: ^(.+):([0-9]+).([0-9]+),([0-9]+).([0-9]+) ([0-9]+) ([0-9]+)$
Makefile:261: recipe for target 'test-job-github.com/vmware/vic/pkg' failed
make: *** [test-job-github.com/vmware/vic/pkg] Error 1
make: *** Waiting for unfinished jobs....
Makefile:261: recipe for target 'test-job-github.com/vmware/vic/cmd' failed
make: *** [test-job-github.com/vmware/vic/cmd] Error 1
ok  	github.com/vmware/vic/lib/imagec	115.164s	coverage: 28.4% of statements
...
```

**Comment by anchal-agrawal** [comment]
Also seen in https://ci.vcna.io/vmware/vic/10671:

```
Running unit tests
# test everything but vendor
Generating coverage data
github.com/vmware/vic/pkg/certificate github.com/vmware/vic/pkg/dio github.com/vmware/vic/pkg/errors github.com/vmware/vic/pkg/fetcher github.com/vmware/vic/pkg/filelock github.com/vmware/vic/pkg/flags github.com/vmware/vic/pkg/fs github.com/vmware/vic/pkg/i18n github.com/vmware/vic/pkg/index github.com/vmware/vic/pkg/ip github.com/vmware/vic/pkg/kvstore github.com/vmware/vic/pkg/log github.com/vmware/vic/pkg/log/syslog github.com/vmware/vic/pkg/logmgr github.com/vmware/vic/pkg/netfilter github.com/vmware/vic/pkg/retry github.com/vmware/vic/pkg/serial github.com/vmware/vic/pkg/telnet github.com/vmware/vic/pkg/trace github.com/vmware/vic/pkg/uid github.com/vmware/vic/pkg/version github.com/vmware/vic/pkg/vsphere/compute github.com/vmware/vic/pkg/vsphere/datastore github.com/vmware/vic/pkg/vsphere/diag github.com/vmware/vic/pkg/vsphere/diagnostic github.com/vmware/vic/pkg/vsphere/disk github.com/vmware/vic/pkg/vsphere/extraconfig github.com/vmware/vic/pkg/vsphere/extraconfig/vmomi github.com/vmware/vic/pkg/vsphere/guest github.com/vmware/vic/pkg/vsphere/session github.com/vmware/vic/pkg/vsphere/simulator github.com/vmware/vic/pkg/vsphere/simulator/esx github.com/vmware/vic/pkg/vsphere/simulator/vc github.com/vmware/vic/pkg/vsphere/sys github.com/vmware/vic/pkg/vsphere/tasks github.com/vmware/vic/pkg/vsphere/test github.com/vmware/vic/pkg/vsphere/test/env github.com/vmware/vic/pkg/vsphere/toolbox github.com/vmware/vic/pkg/vsphere/vm
# github.com/vmware/vic/vendor/github.com/vmware/govmomi/object
vendor/github.com/vmware/govmomi/object/common.go:29: can't find import: "github.com/vmware/vic/vendor/github.com/vmware/govmomi/vim25/types"
Makefile:263: recipe for target 'test-job-github.com/vmware/vic/lib' failed
make: *** [test-job-github.com/vmware/vic/lib] Error 2
make: *** Waiting for unfinished jobs....
```

**Comment by anchal-agrawal** [comment]
https://ci.vcna.io/vmware/vic/10703

**Comment by andrewtchin** [comment]
https://ci.vcna.io/vmware/vic/10852

**Comment by mhagen-vmware** [comment]
Going to close this for now as fixed by #5256 please reopen if we see again.
